test-utils/src/rpc/data/macros.rs

```diff
@@ -0,0 +1,168 @@
+        $( #[$attr:meta] )*
+        $name:ident: $type:ty,
+
+        // The request type (and any doc comments, derives, etc).
+        $( #[$request_attr:meta] )*
+        Request = $request:expr;
+
+        // The response type (and any doc comments, derives, etc).
+        $( #[$response_attr:meta] )*
+        Response = $response:expr;
+    ) => { paste::paste! {
+        #[doc = $crate::rpc::data::macros::define_request_and_response_doc!(
+            "response" => [<$name:upper _RESPONSE>],
+            $monero_daemon_rpc_doc_link,
+        )]
+        ///
+        $( #[$attr] )*
+        ///
+        $( #[$request_attr] )*
+        ///
+        $(
+            #[doc = $crate::rpc::data::macros::define_request_and_response_doc_test!([<$name:upper _REQUEST>], $test)]
+        )?
+        pub const [<$name:upper _REQUEST>]: $type = $request;
+
+        #[doc = $crate::rpc::data::macros::define_request_and_response_doc!(
+            "request" => [<$name:upper _REQUEST>],
+            $monero_daemon_rpc_doc_link,
+        )]
+        ///
+        $( #[$attr] )*
+        ///
+        $( #[$response_attr] )*
+        ///
+        $(
+            #[doc = $crate::rpc::data::macros::define_request_and_response_doc_test!([<$name:upper _RESPONSE>], $test)]
+        )?
+        pub const [<$name:upper _RESPONSE>]: $type = $response;
+    }};
+}
+pub(super) use define_request_and_response;
+
+//---------------------------------------------------------------------------------------------------- define_request_and_response_doc
+/// Generate documentation for the types generated
+/// by the [`define_request_and_response`] macro.
+///
+/// See it for more info on inputs.
+macro_rules! define_request_and_response_doc {
+    (
+        // This labels the last `[request]` or `[response]`
+        // hyperlink in documentation. Input is either:
+        // - "request"
+        // - "response"
+        //
+        // Remember this is linking to the _other_ type,
+        // so if defining a `Request` type, input should
+        // be "response".
+        $request_or_response:literal => $request_or_response_type:ident,
+        $monero_daemon_rpc_doc_link:ident,
+    ) => {
+        concat!(
+            "",
+            "[Documentation](",
+            "https://www.getmonero.org/resources/developer-guides/daemon-rpc.html",
+            "#",
+            stringify!($monero_daemon_rpc_doc_link),
+            "), [",
+            $request_or_response,
+            "](",
+            stringify!($request_or_response_type),
+            ")."
+        )
+    };
+}
+pub(super) use define_request_and_response_doc;
+
+//---------------------------------------------------------------------------------------------------- define_request_and_response_test
+/// Generate documentation for the types generated
+/// by the [`define_request_and_response`] macro.
+///
+/// See it for more info on inputs.
+macro_rules! define_request_and_response_doc_test {
```

Comment:
All `/json_rpc` and other JSON types have tests that assert they are valid JSON.

Tests that assert correct (de)serialization of values will be in the `rpc/` crates.